docs/model_developers/model_debugging.rst

```diff
@@ -0,0 +1,399 @@
+        CONSTRUCTED BY `A(n=template.n)`
+        AT ./entities.cf:11
+    ├── template.n
+    │   SET BY `A(n=template.n)`
+    │   AT ./entities.cf:11
+    │   SUBTREE for template:
+    │       └── x
+    │           SET BY `template = x`
+    │           AT ./entities.cf:9
+    │           └── __config__::A instance
+    │               SET BY `x = A(n=0)`
+    │               AT ./entities.cf:7
+    │               CONSTRUCTED BY `A(n=0)`
+    │               AT ./entities.cf:7
+    │   └── 0
+    │       SET BY `A(n=0)`
+    │       AT ./entities.cf:7
+    └── 1
+        SET BY `y.n = 1`
+        AT ./entities.cf:12
+
+As usual, line 1 states the variable that represents
+the root of the data flow tree. In this case it's the attribute ``n`` of an instance of ``A``. Which instance?
+That is shown in the subtree for that instance on lines 2--4. In this case it's a very simple subtree that shows
+just the construction of the instance and the line number in the configuration model. The tree for the attribute
+starts at line 5. The first branch shows the assignment to ``template.n`` in the constructor for ``y``. Then
+another subtree is shown at lines 8--16, this one more useful. It shows a data flow graph like we're used to
+by now, with ``template`` as the root. Then at line 17 the trace shows the data flow ``template.n <- 0`` referring
+to ``entities.cf:7``. This line doesn't assign to ``template.n`` directly, but it does assign to the instance at the
+end of the subtree for ``template`` (the data that flows to ``template``).
+
+
+Let's have a look at an implementation:
+
+.. code-block:: inmanta
+    :caption: implementation.cf
+    :linenos:
+
+    entity A:
+        number n
+    end
+
+    implement A using i
+
+    implementation i for A:
+        self.n = 42
+    end
+
+    x = A(n = 0)
+
+.. code-block::
+    :caption: data trace for implementation.cf
+    :linenos:
+
+    attribute n on __config__::A instance
+    SUBTREE for __config__::A instance:
+        CONSTRUCTED BY `A(n=0)`
+        AT ./implementation.cf:11
+    ├── 0
+    │   SET BY `A(n=0)`
+    │   AT ./implementation.cf:11
+    └── 42
+        SET BY `self.n = 42`
+        AT ./implementation.cf:8
+        IN IMPLEMENTATION WITH self = __config__::A instance
+            CONSTRUCTED BY `A(n=0)`
+            AT ./implementation.cf:11
+
+The only thing new in this trace can be found at lines 11---13. It highlights that a statement was executed within a dynamic context
+and shows a subtree for the ``self`` variable.
+
+
+And finally, an index:
+
+.. code-block:: inmanta
+    :caption: index.cf
+    :linenos:
+
+    entity A:
+        number n
+        number m
+    end
+
+    index A(n)
+
+    implement A using std::none
+
+    x = A(n = 42, m = 0)
+    x = A(n = 42, m = 1)
+
+.. code-block::
+    :caption: data trace for index.cf
+    :linenos:
+
+    attribute m on __config__::A instance
+    SUBTREE for __config__::A instance:
+        CONSTRUCTED BY `A(n=42,m=0)`
```

Comment:
I miss a `1 SET BY A(n=42,m=1)` ?

Reply:
`SET BY ...` is only output on branches of the tree. This is the root of a subtree. Nothing is assigned to the instance (it's not assignable), it's just constructed.